### DISCUSSION TITLE
Does em++ able to generate standalone wasm with C++ exception?

### DISCUSSION THREAD

**Discussion by harry900831** [emscripten-core/emscripten#19150] [Q&A] [answered]
Hi all,

I'm currently working on adding wasm exception handling proposal support for a runtime. According to my test, it seems that if we generate the wasm with following code and command: ``` em++ -fwasm-exceptions a.cpp -o a.wasm```
, it still depends on an import object called ```(import "env" "__throw_exception_with_stack_trace" (func (;2;) (type 3)))```.

```cpp
#include <stdio.h>

int main() {
  try {
    puts("throw...");
    throw 1;
    puts("(never reached)");
  } catch(...) {
    puts("catch!");
  }
  return 0;
}
```

According to this [page](https://github.com/emscripten-core/emscripten/wiki/WebAssembly-Standalone) (last update on 2019), if I understand correctly, it states that emscripten still can't generate standalone wasm with C++ exceptions.

```
Many APIs that emscripten uses depend on JS or Web APIs, though, like C++ exceptions as mentioned earlier, things like WebGL or WebAudio, etc. 
Calls to those APIs will still be emitted in standalone mode because we have no better alternative; this also gives you the option to add necessary APIs to a custom wasm embedding.
```

Moreover, If I generate the wasm with the command: `em++ -O1 -fwasm-exceptions a.cpp -o a.wasm`, it eliminate the import object ```(import "env" "__throw_exception_with_stack_trace" (func (;2;) (type 3)))```, but the output after running it is "throw..." only.

I would like to ask whether there is any update for standalone wasm that support C++ exception handling. And if capable, how to generate one?

**Comment by kripken** [comment, marked as answer]
`__throw_exception_with_stack_trace` is a debugging helper. That shouldn't be emitted in standalone mode, so we should add an if to remove that. But right now you are getting that since at `-O0` we enable `ASSERTIONS` by default. Building with `-O1` or above (or `-O0 -sASSERTIONS=0`) will avoid that issue.

Aside from that, I'm not aware of any issues. `emcc a.cpp -O1 --profiling -fwasm-exceptions -sSTANDALONE_WASM` on your cpp file emits a standalone wasm with no odd imports, and when I run it using the JS shim in node 16.5 with `--experimental-wasm-eh` it emits
```
throw...
catch!
```
That should work in a standalone wasm VM as well, but I'm not sure which of them support wasm exceptions atm.

**Reply by harry900831** [reply]
Thanks for your reply! My problem is that I misunderstood something related to run the standalone wasm on node. It's all clear now!

**Reply by vchigrin** [reply]
> That shouldn't be emitted in standalone mode, so we should add an if to remove that.

I attempt to fix issue here
https://github.com/emscripten-core/emscripten/pull/24680
